ext/hash/hash.c

```diff
@@ -597,6 +597,108 @@ PHP_FUNCTION(hash_algos)
 }
 /* }}} */
 
+/* {{{ proto string hkdf(string algo, string ikm [, int length = 0, string info = '', string salt = ''])
+RFC5869 HMAC-based key derivation function */
+PHP_FUNCTION(hkdf)
+{
+	zend_string *returnval, *ikm, *algo, *info = NULL, *salt = NULL;
+	zend_long length = 0;
+	char *prk, *computed_salt;
+	unsigned char *digest, *K;
+	int i, rounds;
+	const php_hash_ops *ops;
+	void *context;
+
+	if (zend_parse_parameters(ZEND_NUM_ARGS(), "SS|lSS", &algo, &ikm, &length, &info, &salt) == FAILURE) {
+		return;
+	}
+
+	ops = php_hash_fetch_ops(algo->val, algo->len);
+	if (!ops) {
+		php_error_docref(NULL, E_WARNING, "Unknown hashing algorithm: %s", algo->val, algo->len);
+		RETURN_FALSE;
+	}
+
+	if (ikm->len <= 0) {
+		php_error_docref(NULL, E_WARNING, "Input keying material cannot be empty");
+		RETURN_FALSE;
+	}
+
+	if (length < 0) {
+		php_error_docref(NULL, E_WARNING, "Length must be greater than or equal to 0: " ZEND_LONG_FMT, length);
+		RETURN_FALSE;
+	}
+	else if (length == 0) {
+		length = ops->digest_size;
+	}
+	else if (length > ops->digest_size * 255) {
+		php_error_docref(NULL, E_WARNING, "Length must be less than or equal to %d: " ZEND_LONG_FMT, ops->digest_size * 255, length);
+		RETURN_FALSE;
+	}
+
+	if (salt != NULL && salt->len > 0) {
+		computed_salt = emalloc(salt->len);
+		memcpy(computed_salt, salt->val, salt->len);
+	}
+	else {
+		computed_salt = emalloc(ops->digest_size);
+		for (i = 0; i < ops->digest_size; i++)
+		{
+			computed_salt[i] = 0x00;
+		}
+	}
+
+	context = emalloc(ops->context_size);
+
+	// Extract
+	ops->hash_init(context);
+	K = emalloc(ops->block_size);
+	php_hash_hmac_prep_key(K, ops, context, computed_salt, (salt != NULL && salt->len ? salt->len : ops->digest_size));
+	prk = safe_emalloc(ops->digest_size, ops->digest_size, 0);
+	php_hash_hmac_round(prk, ops, context, K, ikm->val, ikm->len);
+	php_hash_string_xor_char(K, K, 0x6A, ops->block_size);
+	php_hash_hmac_round(prk, ops, context, K, prk, ops->digest_size);
+	ZEND_SECURE_ZERO(K, ops->block_size);
+	efree(computed_salt);
+
+	// Expand
+	returnval = zend_string_alloc(length, 0);
+	digest = emalloc(ops->digest_size);
+	for (i = 1, rounds = ceil((float) length / (float) ops->digest_size); i <= rounds; i++) {
```

Comment:
Using floats here may result in rounding errors for large lengths. Please use [ceil integer division](http://stackoverflow.com/questions/17944/how-to-round-up-the-result-of-integer-division) instead.